### ISSUE TITLE
Question Optics: Setter (.~) needs the property to be mutable?

### ISSUE THREAD

**Issue by amadeu01** [pointfreeco/swift-prelude#130]
That works fine:

```
struct User {
  var id: Int
  var name: String
}

let user = User(id: 1, name: "Stephen")

let newUser = user
    |> \.id .~ 2
    |> \.name .~ "Amadeu"
```

However, that:

```
struct User {
  let id: Int. // Replaced var by let
  var name: String
}
```

Does not work.

```
let newUser = user
    |> \User.id .~ 2
    |> \User.name .~ "Amadeu"

consecutive statements on a line must be separated by ';'
    |> \User.id .~ 2
```

```

let newUser = user
    |> \.id .~ 2
    |> \.name .~ "Amadeu"


error: type of expression is ambiguous without more context
    |> \.id .~ 2
       ^~~~
```

**Comment by seanwoodward** [comment]
In short, yes, the properties need to be mutable. 

This is tested in [OpticsTests.swift](https://github.com/pointfreeco/swift-prelude/blob/7056dd00e89cda09ee4b64bc1ab4f6f9c013a61c/Tests/OpticsTests/OpticsTests.swift#L84) with [an Episode and a User](https://github.com/pointfreeco/swift-prelude/blob/7056dd00e89cda09ee4b64bc1ab4f6f9c013a61c/Tests/OpticsTests/OpticsTests.swift#L11-L37). The properties on these structs are declared with `private(set) var`. I don't know why the test passes because `private(set)` should restrict write access to the struct or it's extensions in the same file. Changing the declaration to `fileprivate(set) var` will allow the `KeyPath` to be recognized as a `WritableKeyPath`. In the case of your `\User.id`, any property declared with `let` is read only, so the key path will not be a `WritableKeyPath`. You could also use `internal(set)` for module level access to write to the property. Note, the setter does not change the value of the original user, but still requires write access to the property to update the copy.

@stephencelis, @mbrandonw can you provide some insight as to why the tests pass?

**Comment by mbrandonw** [comment]
@seanwoodward This is because swift-prelude is still on Swift 4.2 :/ We haven't yet merged #126, and it was in Swift 5 that a bug was fixed that made `private(set) var` no longer writeable.

**Comment by seanwoodward** [comment]
@mbrandonw I thought that was the case, but did not see a reference to the fix in the release notes for Swift 5.
